### PR TITLE
Issues/677 dont offer contacts for whom we dont have contacts

### PR DESCRIPTION
**Pull request by lfalvarez**
This PR filters the persons without contacts in the message creating part in the 'who' step.

### CODE REVIEW

File: nuntium/models.py [hunk 1]
```diff
@@ -55,6 +55,13 @@ class WriteItInstance(models.Model):
         through='Membership')
     owner = models.ForeignKey(User, related_name="writeitinstances")
 
+    def add_person(self, person):
+        Membership.objects.get_or_create(writeitinstance=self, person=person)
+
+    @property
+    def persons_with_contacts(self):
+        return self.persons.filter(contact__isnull=False)
```
Comment (duncanparkes) [inline nuntium/models.py]: What happens if a Person has multiple Contacts? I think you're going to get a row per Contact/Person combination. Is that what you want?

File: nuntium/views.py [hunk 2]
```diff
@@ -133,6 +132,7 @@ def get_context_data(self, form, **kwargs):
             context['persons'] = self.writeitinstance.persons.all()
         if self.steps.current == 'preview':
             context['message'] = self.get_form_values()
+            context['message']['people'] = context['message']['persons']
```
Comment (duncanparkes) [inline nuntium/views.py]: I wonder here if we should just pick one of 'persons' and 'people' and stick with it.

File: nuntium/tests/message_creation_tests.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        response = self.client.get(url)
+        who_form = response.context['form']
+        self.assertIsInstance(who_form, WhoForm)
+        self.assertIn(self.person1, who_form.fields['persons'].queryset)
+        self.assertNotIn(self.person2, who_form.fields['persons'].queryset)
+
+        self.assertIn('writeitinstance', response.context)
+        self.assertEquals(response.context['writeitinstance'], self.writeitinstance)
+        recipient1 = self.writeitinstance.persons.first()
```
Comment (duncanparkes) [inline nuntium/tests/message_creation_tests.py]: If there are multiple people in `self.writeitinstance.persons`, I'd be more comfortable with choosing a particular one with `.get()`.

File: nuntium/tests/message_creation_tests.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        response = self.client.post(url, data={
+            'who-persons': [recipient1.id],
+            'write_message_view-current_step': 'who'
+            })
+        self.assertEquals(response.status_code, 302)
```
Comment (duncanparkes) [inline nuntium/tests/message_creation_tests.py]: This bit feels like a case for `assertRedirects`, which will handle the status code and the expected url in one go.

File: nuntium/tests/message_creation_tests.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        response = self.client.get(url2)
+        self.assertIsInstance(response.context['form'], DraftForm)
+        response = self.client.post(url2, data={
+            'draft-subject': 'This is the subjectand is very specific',
+            'draft-content': "This is the content",
+            'draft-author_name': "Feli",
+            'draft-author_email': 'email@mail.com',
+            'write_message_view-current_step': 'draft'
+            })
+        self.assertEquals(response.status_code, 302)
```
Comment (duncanparkes) [inline nuntium/tests/message_creation_tests.py]: Ditto here - `assertRedirects` is your friend.

File: nuntium/tests/message_creation_tests.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        url3 = reverse('write_message_step',
+            subdomain=self.writeitinstance.slug,
+            kwargs={'step': 'preview'})
+
+        self.assertEquals(response.url, url3)
+        response = self.client.get(url3)
+        self.assertEquals(response.context['message']['persons'][0].id, self.person1.id)
+        # in this step I'm oing to replicate the content of
```
Comment (duncanparkes) [inline nuntium/tests/message_creation_tests.py]: Typo s/oing/going/.

File: nuntium/tests/writeitinstances_test.py [hunk 7]
```diff
@@ -141,6 +141,38 @@ def test_it_has_a_pulling_from_popit_status(self):
...
+        self.person2.contact_set.all()
+
+        self.assertTrue(self.person2.contact_set.all())
+        # self.person1 does not have contacts
+        # self.person2 has contacts
+        self.writeitinstance.add_person(self.person1)
+        self.writeitinstance.add_person(self.person2)
+
+        self.assertNotIn(self.person1,
+            self.writeitinstance.persons_with_contacts.all())
```
Comment (duncanparkes) [inline nuntium/tests/writeitinstances_test.py]: You don't really need the `.all()` on these.

File: nuntium/views.py [hunk 8]
```diff
@@ -90,8 +90,7 @@ def get_template_names(self):
 
     def get_form_kwargs(self, step):
         if step == 'who':
-            # FIXME: Do we actually want .all() here?
-            return {'persons_queryset': self.writeitinstance.persons.all()}
+            return {'persons_queryset': self.writeitinstance.persons_with_contacts.all()}
```
Comment (duncanparkes) [inline nuntium/views.py]: You don't need the `.all()` here either - you've already done a filter.

File: nuntium/tests/message_creation_tests.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+class MessageCreationTestCase(TestCase):
+    def setUp(self):
+        super(MessageCreationTestCase, self).setUp()
+        self.writeitinstance = WriteItInstance.objects.get(id=1)
+        self.person1 = Person.objects.get(id=1)
+        self.person2 = Person.objects.get(id=2)
+        self.writeitinstance.add_person(self.person1)
+        self.writeitinstance.add_person(self.person2)
+
+    def test_step_who_doesn_have_persons_without_contacts_self_client(self):
```
Comment (duncanparkes) [inline nuntium/tests/message_creation_tests.py]: I don't understand what this method name and comment are trying to say, I'm afraid.

**Comment by duncanparkes** [comment]
Putting aside minor things, the only real issue is what happens if a person has more than one contact? Previously, you would have got a list with only one entry per person in it. I now think you'll get an entry per contact. I suspect this is not what you want, but I'm not sure. Perhaps pick the first email address for a person?

File: nuntium/models.py (hunk 1 above)
Comment (lfalvarez) [inline nuntium/models.py]: Ough! I hadn't seen that one! thanks!

Comment (lfalvarez) [inline nuntium/models.py]: Actually this is a deeper thing, a person could have contacts related to another instance and therefore would be appearing in this queryset too! thanks this was a bug

File: nuntium/views.py (hunk 2 above)
Comment (lfalvarez) [inline nuntium/views.py]: I think that that the problem here is unify the naming related to ```Message.persons``` and ```Message.people``` for which I'm opening a new ticket.

**Comment by chrismytton** [comment]
:rocket: :+1:

**Comment by tmtmtmtm** [comment]
Closes #677

**Pull request merged**